Tests/kaas/k8s-version-recency/k8s-version-recency-check.py

```diff
@@ -0,0 +1,339 @@
+                    vi = K8sVersionInfo()
+                    if 'lessThanOrEqual' in v.keys():
+                        vi.extract_version(v['lessThanOrEqual'].strip("v"))
+                        vi.patch += 1
+                    elif 'lessThan' in v.keys():
+                        vi.extract_version(v['lessThan'].strip("v"))
+                    if vi not in cfvs:
+                        cfvs.append(vi)
+        except KeyError as e:
+            logger.debug(f"They key {e} couldn't be found in the CVE json data for CVE {cveid}.")
+            pass
+
+    return cfvs
+
+
+async def get_k8s_cluster_version(kubeconfig):
+
+    await kubernetes_asyncio.config.load_kube_config(kubeconfig)
+
+    async with kubernetes_asyncio.client.ApiClient() as api:
+        version_api = kubernetes_asyncio.client.VersionApi(api)
+        ret = await version_api.get_code()
+
+        version = K8sVersionInfo()
+        version.extract_version(ret.git_version.strip("v"))
+        version.date = datetime.datetime.strptime(ret.build_date, '%Y-%m-%dT%H:%M:%SZ')
+
+        with open(kubeconfig, "r") as fd:
+            cluster_config = yaml.safe_load(fd)
+            fd.close()
+
+        return version, cluster_config['contexts'][0]['name']
+
+
+def get_k8s_versions(token, cve_versions=None):
+
+    k8s_versions = []
+
+    github_headers = {
+        "Accept": "application/vnd.github+json",
+        "Authorization": f"Bearer {token}",
+        "X-GitHub-Api-Version": "2022-11-28"
+    }
+
+    # Request latest version
+    resp = requests.get("https://api.github.com/repos/kubernetes/kubernetes/releases/latest",
+                        headers=github_headers).json()
+
+    # Latest version
+    lv = K8sVersionInfo()
+    lv.extract_version(resp['tag_name'].strip("v"))
+    lv.date = datetime.datetime.strptime(resp['published_at'], '%Y-%m-%dT%H:%M:%SZ')
+
+    # Request the latest 100 version (the next are not needed, since these versions are too old)
+    resp = requests.get("https://api.github.com/repos/kubernetes/kubernetes/releases?per_page=100",
+                        headers=github_headers).json()
+
+    # Iterate all versions until the first patch of the previous latest version
+    for i in range(0, len(resp)):
+        v = K8sVersionInfo()
+        v.extract_version(resp[i]['tag_name'].split("-")[0].strip("v"))
+        v.date = datetime.datetime.strptime(resp[i]['published_at'], '%Y-%m-%dT%H:%M:%SZ')
+
+        if resp[i]['draft'] is False and resp[i]['prerelease'] is False:
+            k8s_versions.append(v)
+
+        # Stop adding new version if the version we added is the previous latest minor versions first patch,
+        # since it is most of the time around 3-6 months until a new version comes out or an old one goes into EOL
+        if v.check_for_version(lv.major, (lv.minor-1), 0):
+            break
+
+    # Filter the versions depending on the usage times set by the standard
+    for i in range(0, len(k8s_versions)):
```

Comment:
So this loop seems to replace certain entries in `k8s_versions` with `None`, only to then use list comprehension to obtain the list of the other entries. I think there are two more preferable ways to achieve this aim:
* Ideally, use list comprehension directly: `filtered = [v for v in k8s_versions if is_admissible(v)]` where you would extract the bulk of the loop's body into the new function `is_admissible`
* Or, at least use `enumerate` (`for i, v in enumerate(k8s_versions):`) -- the use of `k8s_versions.__getitem__(i)` instead of `k8s_versions[i]` is very baffling and (if really necessary) needs to be accompanied with a comment, but if `__getitem__` isn't necessary, then just use `enumerate`.

To summarize: If at all possible, try to do the list comprehension directly.

Reply:
Separated the comprehension in a new function.